File: scripts/fetch_part_images.py
#!/usr/bin/env python3
"""
Download Beyblade X part images from beyblade.fandom.com.
Skips images that already exist unless --force is given.
Updates src/data/parts-overrides.json with new filenames.

Usage:
    python scripts/fetch_part_images.py           # download missing only
    python scripts/fetch_part_images.py --force   # re-download all, delete old
    python scripts/fetch_part_images.py --dry-run # preview changes
"""

import json
import os
import sys
import time
import urllib.parse
import urllib.request
from pathlib import Path

ROOT = Path(__file__).parent.parent
IMG_DIR = ROOT / "public" / "images"
OVERRIDES_PATH = ROOT / "src" / "data" / "parts-overrides.json"

# Never delete these
PRESERVE = {
    "attack.png", "balance.png", "defense.png", "stamina.png",
    "gcash-qr.jpg", "left-spin.png", "right-spin.png",
    "BladeUnknown.svg", "RatchetBitOperate.webp",
}

HEADERS = {"User-Agent": "Mozilla/5.0"}
API = "https://beyblade.fandom.com/api.php"
WIKI_STATIC = "https://static.wikia.nocookie.net"

DRY_RUN = "--dry-run" in sys.argv
FORCE = "--force" in sys.argv


def wiki_titles_for_overrides(overrides: dict) -> list[tuple]:
    """
    Build list of (category, key, wiki_page_title) for all parts with images.
    Returns only parts that have non-BladeUnknown images currently.
    """
    tasks = []

    # --- Blades ---
    blade_special = {
        "L-DRAGO": "Blade_-_Lightning_L-Drago_(Upper_Type)",
        "CROC CRUNCH": "Blade_-_Bite_Croc",
        "BEARSCRATCH": "BearScratch_5-60F",
        "SAMURAI SABER": "Blade_-_SamuraiSaber",
        "TUSK MAMMOTH": "Blade_-_Tusk_Mammoth",
        "YELL KONG": "Blade_-_Yell_Kong",
        # Wiki uses reversed/different names for these parts
        "PTERASWING": "Blade_-_Talon_Ptera",
        "SHINOBIKNIFE": "Blade_-_Knife_Shinobi",
        "SHARKGILL": "Blade_-_Gill_Shark",
        "GOATTACKLE": "Blade_-_Tackle_Goat",
        "SAMURAISTEEL": "Steel_Samurai_4-80T",
        "WYVERNHOVER": "Blade_-_Hover_Wyvern",
        "ROCKLEONE": "Blade_-_Rock_Leone",
        "TYRANNOROAR": "Blade_-_Roar_Tyranno",
    }
    for key, entry in overrides.get("blades", {}).items():
        img = entry.get("image", "")
        if not img or img == "BladeUnknown.svg":
            continue
        if key in blade_special:
            wiki_title = blade_special[key]
        else:
            name = entry.get("name", key)
            wiki_title = "Blade_-_" + name.replace(" ", "")
        tasks.append(("blades", key, wiki_title))

    # --- Main Blades ---
    for key, entry in overrides.get("mainBlades", {}).items():
        img = entry.get("image", "")
        if not img or img == "BladeUnknown.svg":
            continue
        name = entry.get("name", key.title())
        wiki_title = "Main_Blade_-_" + name
        tasks.append(("mainBlades", key, wiki_title))

    # --- Assist Blades ---
    for key, entry in overrides.get("assistBlades", {}).items():
        img = entry.get("image", "")
        if not img or img == "BladeUnknown.svg":
            continue
        name = entry.get("name", key)
        wiki_title = "Assist_Blade_-_" + name
        tasks.append(("assistBlades", key, wiki_title))

    # --- Lock Chips ---
    for key, entry in overrides.get("lockChips", {}).items():
        img = entry.get("image", "")
        if not img or img == "BladeUnknown.svg":
            continue
        name = entry.get("name", key.title())
        wiki_title = "Lock_Chip_-_" + name
        tasks.append(("lockChips", key, wiki_title))

    return tasks


def batch_query_images(titles: list[str]) -> dict[str, dict]:
    """Query wiki API for page images in batches of 50. Returns {title: {pageimage, url}}."""
    results = {}
    batch_size = 50
    for i in range(0, len(titles), batch_size):
        batch = titles[i : i + batch_size]
        joined = "%7C".join(urllib.parse.quote(t, safe="_()-") for t in batch)
        url = f"{API}?action=query&titles={joined}&prop=pageimages&pithumbsize=500&format=json"
        req = urllib.request.Request(url, headers=HEADERS)
        with urllib.request.urlopen(req) as r:
            data = json.load(r)

        redirects = {v: k for k, v in data.get("query", {}).get("redirects", {}).items() if isinstance(k, str)}
        normalized = {v: k for v, k in [(n.get("to"), n.get("from")) for n in data.get("query", {}).get("normalized", [])]}

        for _pid, page in data["query"]["pages"].items():
            title = page["title"]
            # MediaWiki normalizes underscores to spaces; store with underscores to match request titles
            title_key = title.replace(" ", "_")
            pageimage = page.get("pageimage")
            thumb_url = page.get("thumbnail", {}).get("source", "")
            if pageimage and thumb_url:
                clean_url = thumb_url.split("/scale-to-width-down")[0].split("/revision/latest")[0] + "/revision/latest"
                results[title_key] = {"pageimage": pageimage, "url": clean_url}
            else:
                results[title_key] = {"pageimage": None, "url": None}
        time.sleep(0.2)
    return results


def download_image(url: str, dest_path: Path) -> bool:
    """Download image from url to dest_path. Returns True on success."""
    try:
        req = urllib.request.Request(url, headers=HEADERS)
        with urllib.request.urlopen(req) as r, open(dest_path, "wb") as f:
            f.write(r.read())
        return True
    except Exception as e:
        print(f"    ERROR downloading {url}: {e}")
        return False


def ext_from_filename(filename: str) -> str:
    return Path(filename).suffix.lower()


def main():
    with open(OVERRIDES_PATH, encoding="utf-8") as f:
        overrides = json.load(f)

    tasks = wiki_titles_for_overrides(overrides)
    print(f"Found {len(tasks)} parts with images to update")

    # Query wiki for all titles
    all_titles = [t for _, _, t in tasks]
    print(f"Querying wiki API for {len(all_titles)} page titles...")
    image_data = batch_query_images(all_titles)

    # Report which titles had no image
    missing = [(cat, key, title) for cat, key, title in tasks if not image_data.get(title, {}).get("pageimage")]
    if missing:
        print(f"\n{len(missing)} parts had no wiki image:")
        for cat, key, title in missing:
            print(f"  [{cat}] {key} -> {title}")

    # Build download plan: (category, key, wiki_pageimage, download_url, new_filename)
    plan = []
    for cat, key, wiki_title in tasks:
        info = image_data.get(wiki_title, {})
        if not info.get("pageimage"):
            continue
        pageimage = info["pageimage"]
        url = info["url"]
        ext = ext_from_filename(pageimage)
        # Use wiki's canonical filename (pageimage field)
        new_filename = pageimage
        plan.append((cat, key, wiki_title, url, new_filename))

    print(f"\nWill download {len(plan)} images")

    # Filter plan: skip already-downloaded images unless --force
    if not FORCE:
        skipped = [(cat, key, wt, url, fn) for cat, key, wt, url, fn in plan if (IMG_DIR / fn).exists()]
        plan = [(cat, key, wt, url, fn) for cat, key, wt, url, fn in plan if not (IMG_DIR / fn).exists()]
        if skipped:
            print(f"Skipping {len(skipped)} already-downloaded images (use --force to re-download)")

    if DRY_RUN:
        print("\n--- DRY RUN ---")
        for cat, key, wiki_title, url, new_filename in plan:
            old_filename = overrides[cat][key].get("image", "")
            status = "(new)" if old_filename == new_filename else f"{old_filename!r} ->"
            print(f"  [{cat}] {key}: {status} {new_filename!r}")
        return

    # --force: delete old part images before downloading fresh ones
    if FORCE:
        new_filenames = {item[4] for item in plan}
        images_to_delete = [
            fname for fname in os.listdir(IMG_DIR)
            if fname not in PRESERVE and fname not in new_filenames
        ]
        print(f"\nDeleting {len(images_to_delete)} old images...")
        for fname in sorted(images_to_delete):
            fpath = IMG_DIR / fname
            if fpath.exists():
                os.remove(fpath)
                print(f"  deleted: {fname}")

    # Download images and update overrides
    print(f"\nDownloading {len(plan)} images...")
    updated_overrides = json.loads(json.dumps(overrides))  # deep copy

    for cat, key, wiki_title, url, new_filename in plan:
        dest = IMG_DIR / new_filename
        print(f"  [{cat}] {key}: {new_filename}", end=" ... ", flush=True)
        if download_image(url, dest):
            print("OK")
            updated_overrides[cat][key]["image"] = new_filename
        else:
            print("FAILED")
        time.sleep(0.1)

    # Save updated overrides
    with open(OVERRIDES_PATH, "w", encoding="utf-8") as f:
        json.dump(updated_overrides, f, indent=2, ensure_ascii=False)
        f.write("\n")
    print(f"\nUpdated: {OVERRIDES_PATH}")
    print("Run: python scripts/generate_parts.py  to regenerate beyparts.js")


if __name__ == "__main__":
    main()

File: scripts/fetch_part_images.js
#!/usr/bin/env node
/**
 * Download Beyblade X part images from beyblade.fandom.com.
 * Skips images that already exist unless --force is given.
 * Updates src/data/parts-overrides.json with new filenames.
 *
 * Usage:
 *   node scripts/fetch_part_images.js           # download missing only
 *   node scripts/fetch_part_images.js --force   # re-download all, delete old
 *   node scripts/fetch_part_images.js --dry-run # preview changes
 */

import { createWriteStream, existsSync, readdirSync, readFileSync, renameSync, rmSync } from "fs";
import { readFile, writeFile } from "fs/promises";
import { pipeline } from "stream/promises";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const IMG_DIR = path.join(ROOT, "public", "images");
const OVERRIDES_PATH = path.join(ROOT, "src", "data", "parts-overrides.json");

const PRESERVE = new Set([
  "attack.png", "balance.png", "defense.png", "stamina.png",
  "gcash-qr.jpg", "left-spin.png", "right-spin.png",
  "BladeUnknown.svg", "RatchetBitOperate.webp",
]);

const HEADERS = { "User-Agent": "Mozilla/5.0" };
const API = "https://beyblade.fandom.com/api.php";

const DRY_RUN = process.argv.includes("--dry-run");
const FORCE = process.argv.includes("--force");

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// eslint-disable-next-line no-unused-vars
async function apiGet(params) {
  const qs = new URLSearchParams(params).toString();
  const res = await fetch(`${API}?${qs}`, { headers: HEADERS });
  return res.json();
}

function isWebP(filePath) {
  try {
    const buf = readFileSync(filePath);
    return buf.length >= 12 && buf.toString('ascii', 8, 12) === 'WEBP';
  } catch {
    return false;
  }
}

async function downloadImage(url, destPath) {
  try {
    const res = await fetch(url, { headers: HEADERS });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    await pipeline(res.body, createWriteStream(destPath));
    // Wiki sometimes serves WebP with wrong extension; fix it
    if ((destPath.endsWith('.png') || destPath.endsWith('.jpeg') || destPath.endsWith('.jpg')) && isWebP(destPath)) {
      const webpPath = destPath.replace(/\.(png|jpeg|jpg)$/, '.webp');
      renameSync(destPath, webpPath);
      return webpPath;
    }
    return destPath;
  } catch (e) {
    console.log(`    ERROR downloading ${url}: ${e.message}`);
    return false;
  }
}

function cleanImageUrl(thumbUrl) {
  return (
    thumbUrl.split("/scale-to-width-down")[0].split("/revision/latest")[0] +
    "/revision/latest"
  );
}

function wikiTitlesForOverrides(overrides) {
  const tasks = [];

  const bladeSpecial = {
    "LIGHTNING L-DRAGO": "Blade_-_Lightning_L-Drago_(Upper_Type)",
    "L-DRAGO": "Blade_-_Lightning_L-Drago_(Upper_Type)",
    "CROCOCRUNCH": "Blade_-_Bite_Croc",
    "BEARSCRATCH": "BearScratch_5-60F",
    "MAMMOTHTUSK": "Blade_-_Tusk_Mammoth",
    "YELL KONG": "Blade_-_Yell_Kong",
    "PHOENIXFLARE": "Main_Blade_-_Flare",
    "PTERASWING": "Blade_-_Talon_Ptera",
    "SHINOBIKNIFE": "Blade_-_Knife_Shinobi",
    "SHARKGILL": "Blade_-_Gill_Shark",
    "GOATTACKLE": "Blade_-_Tackle_Goat",
    "SAMURAISTEEL": "Steel_Samurai_4-80T",
    "WYVERNHOVER": "Blade_-_Hover_Wyvern",
    "ROCKLEONE": "Blade_-_Rock_Leone",
    "TYRANNOROAR": "Blade_-_Roar_Tyranno",
    "STORMSPRIGGAN": "Blade_-_StormSpriggan",
    "GLORYVALKYRIE": "Ratchet-Integrated_Blade_-_GloryValkyrie",
    "BULLETGRIFFON": "Ratchet-Integrated_Blade_-_BulletGriffon",
  };

  const mainBladeSpecial = {
    "RIGGLE": "Main_Blade_-_Wriggle",
    "ANTLERS": "BucksAntlers_B2-60D",
  };

  for (const [key, entry] of Object.entries(overrides.blades ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const wikiTitle =
      bladeSpecial[key] ??
      "Blade_-_" + (entry.name ?? key).replace(/\s+/g, "");
    tasks.push(["blades", key, wikiTitle]);
  }

  for (const [key, entry] of Object.entries(overrides.mainBlades ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const name = entry.name ?? key.replace(/\b\w/g, (c) => c.toUpperCase());
    const wikiTitle = mainBladeSpecial[key] ?? `Main_Blade_-_${name}`;
    tasks.push(["mainBlades", key, wikiTitle]);
  }

  for (const [key, entry] of Object.entries(overrides.metalBlades ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const name = entry.name ?? key.replace(/\b\w/g, (c) => c.toUpperCase());
    tasks.push(["metalBlades", key, `Metal_Blade_-_${name}`]);
  }

  const assistBladeSpecial = {
    "G": "Assist_Blade_-_Gravity",
  };

  for (const [key, entry] of Object.entries(overrides.assistBlades ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const name = entry.name ?? key;
    const wikiTitle = assistBladeSpecial[key] ?? `Assist_Blade_-_${name}`;
    tasks.push(["assistBlades", key, wikiTitle]);
  }

  const lockChipSpecial = {
    "BUCKS": "Lock_Chip_-_Stag",
  };

  for (const [key, entry] of Object.entries(overrides.lockChips ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const name = entry.name ?? key.replace(/\b\w/g, (c) => c.toUpperCase());
    const wikiTitle = lockChipSpecial[key] ?? `Lock_Chip_-_${name}`;
    tasks.push(["lockChips", key, wikiTitle]);
  }

  const bitSpecial = {};

  for (const [key, entry] of Object.entries(overrides.bits ?? {})) {
    const img = entry.image ?? "";
    if (!img) continue;
    const name = (entry.name ?? key).replace(/\s+/g, "_");
    const wikiTitle = bitSpecial[key] ?? `Bit_-_${name}`;
    tasks.push(["bits", key, wikiTitle]);
  }

  return tasks;
}

async function batchQueryImages(titles) {
  const results = {};
  const batchSize = 50;
  for (let i = 0; i < titles.length; i += batchSize) {
    const batch = titles.slice(i, i + batchSize);
    const joined = batch
      .map((t) => encodeURIComponent(t).replace(/%20/g, "_"))
      .join("%7C");
    const url = `${API}?action=query&titles=${joined}&prop=pageimages&pithumbsize=500&format=json`;
    const res = await fetch(url, { headers: HEADERS });
    const data = await res.json();

    for (const [, page] of Object.entries(data.query.pages)) {
      const titleKey = page.title.replace(/ /g, "_");
      const pageimage = page.pageimage ?? null;
      const thumbUrl = page.thumbnail?.source ?? "";
      if (pageimage && thumbUrl) {
        results[titleKey] = { pageimage, url: cleanImageUrl(thumbUrl) };
      } else {
        results[titleKey] = { pageimage: null, url: null };
      }
    }
    await sleep(200);
  }
  return results;
}

async function main() {
  const overrides = JSON.parse(await readFile(OVERRIDES_PATH, "utf-8"));

  const tasks = wikiTitlesForOverrides(overrides);
  console.log(`Found ${tasks.length} parts with images to update`);

  const allTitles = tasks.map(([, , t]) => t);
  console.log(`Querying wiki API for ${allTitles.length} page titles...`);
  const imageData = await batchQueryImages(allTitles);

  const missing = tasks.filter(
    ([, , title]) => !imageData[title]?.pageimage
  );
  if (missing.length) {
    console.log(`\n${missing.length} parts had no wiki image:`);
    for (const [cat, key, title] of missing) {
      console.log(`  [${cat}] ${key} -> ${title}`);
    }
  }

  let plan = [];
  for (const [cat, key, wikiTitle] of tasks) {
    const info = imageData[wikiTitle] ?? {};
    if (!info.pageimage) continue;
    plan.push([cat, key, wikiTitle, info.url, info.pageimage]);
  }

  console.log(`\nWill download ${plan.length} images`);

  if (!FORCE) {
    const skipped = plan.filter(([, , , , fn]) =>
      existsSync(path.join(IMG_DIR, fn))
    );
    plan = plan.filter(
      ([, , , , fn]) => !existsSync(path.join(IMG_DIR, fn))
    );
    if (skipped.length) {
      console.log(
        `Skipping ${skipped.length} already-downloaded images (use --force to re-download)`
      );
    }
  }

  if (DRY_RUN) {
    console.log("\n--- DRY RUN ---");
    for (const [cat, key, , , newFilename] of plan) {
      const oldFilename = overrides[cat]?.[key]?.image ?? "";
      const status =
        oldFilename === newFilename ? "(new)" : `${JSON.stringify(oldFilename)} ->`;
      console.log(`  [${cat}] ${key}: ${status} ${JSON.stringify(newFilename)}`);
    }
    return;
  }

  if (FORCE) {
    const newFilenames = new Set(plan.map(([, , , , fn]) => fn));
    const toDelete = readdirSync(IMG_DIR).filter(
      (f) => !PRESERVE.has(f) && !newFilenames.has(f)
    );
    console.log(`\nDeleting ${toDelete.length} old images...`);
    for (const fname of toDelete.sort()) {
      const fpath = path.join(IMG_DIR, fname);
      if (existsSync(fpath)) {
        rmSync(fpath);
        console.log(`  deleted: ${fname}`);
      }
    }
  }

  console.log(`\nDownloading ${plan.length} images...`);
  const updatedOverrides = JSON.parse(JSON.stringify(overrides));

  for (const [cat, key, , url, newFilename] of plan) {
    const dest = path.join(IMG_DIR, newFilename);
    process.stdout.write(`  [${cat}] ${key}: ${newFilename} ... `);
    const result = await downloadImage(url, dest);
    const ok = result !== false;
    const actualFilename = ok ? path.basename(result) : newFilename;
    console.log(ok ? "OK" : "FAILED");
    if (ok) updatedOverrides[cat][key].image = actualFilename;
    await sleep(100);
  }

  await writeFile(
    OVERRIDES_PATH,
    JSON.stringify(updatedOverrides, null, 2) + "\n",
    "utf-8"
  );
  console.log(`\nUpdated: ${OVERRIDES_PATH}`);
  console.log("Run: node scripts/generate_parts.js  to regenerate beyparts.js");
}

main().catch((e) => {
  console.error(e.message);
  process.exit(1);
});
